### DISCUSSION TITLE
8. TCP에서 패킷(세그먼트)이 유실되었는지 어떻게 판단하나요? 유실되면 어떻게 하나요?

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#13] [General]
TCP에서 패킷(세그먼트)이 유실되었는지 어떻게 판단하나요? 유실되면 어떻게 하나요?

**[키워드]**
- TCP 오류 제어
- TCP 재전송(ARQ)
- TCP에서 패킷 유실(손실) 여부 판단
- TCP에서 패킷 유실(손실) 시 처리 방법

**Comment by cutehumanS2** [comment]
TCP는 **중복된 ACK 세그먼트를 3번 이상 수신**하거나 **타임아웃이 발생했을 때** 세그먼트 유실을 감지합니다. 
유실이 감지되면 해당 세그먼트를 **재전송**하며, TCP의 재전송 기법(ARQ)에는 **Stop-and-Wait, Go-Back-N, Selective Repeat**이 있습니다.

**Comment by cutehumanS2** [comment]
### TCP가 오류를 검출하고 세그먼트를 재전송하는 상황

**1) 중복된 ACK 세그먼트를 수신했을 때**
수신 호스트 측이 받은 세그먼트의 순서 번호 중 일부가 누락되면 중복된 ACK 세그먼트를 전송하게 됩니다. 송신 호스트가 3개의 중복 ACK를 받으면 세그먼트가 유실되었음을 판단하고 해당 세그먼트를 재전송합니다. 

**2) 타임아웃이 발생했을 때**
TCP 세그먼트를 송신하는 호스트는 모두 재전송 타이머(RTO, Retransmission Time Out)라는 값을 유지합니다. 호스트가 세그먼트를 전송할 때마다 재전송 타이머를 시작하게 되는데,  이 타이머의 카운트다운이 끝난 상황을 타임아웃이라고 합니다. 타임아웃이 발생할 때까지 ACK 세그먼트를 받지 못하면 세그먼트가 상대 호스트에게 정상적으로 도착하지 않았다고 간주하여 세그먼트를 재전송합니다. 

### ARQ: 재전송 기법

수신 호스트의 답변(ACK)과 타임아웃 발생을 토대로 문제를 진단하고, 문제가 생긴 메시지를 재전송함으로써 신뢰성을 확보하는 방식을 ARQ(Automatic Repeat reQuest, 자동 재전송 요구)라고 합니다. 

대표적인 ARQ 방식으로는 Stop-and-Wait ARQ과 Go-Back-N ARQ 그리고 Selective Repeat ARQ가 있습니다.

*+) 전송 계층의 TCP는 ARQ를 사용하는 대표적인 프로토콜이지만, ARQ 자체는 전송 계층만의 기술은 아닙니다.* 

**1) Stop-and-Wait(정지-대기) ARQ**

- 제대로 전달했음을 확인하기 전까지는 새로운 메시지를 보내지 않는 방식입니다. 단순하지만 높은 신뢰성을 보장하는 방식입니다.
- 송신 호스트 입장에서 확인 응답을 받기 전까지는 다음 전송을 할 수 있어도 하지 못하므로 네트워크의 이용 효율이 낮아질 수 있다는 단점이 있습니다.
- 오늘날 인터넷 환경의 TCP에서는 특별한 경우가 아닌 이상 Stop-and-Wait ARQ를 잘 사용하지 않습니다.

![image](https://github.com/user-attachments/assets/fba43d49-4fb7-4326-9f8a-48c68bd3f377)

**2) Go-Back-N ARQ**

- Stop-and-Wait과 달리 파이프라이닝 방식을 활용해 여러 세그먼트를 전송하고, 도중에 누락된 세그먼트가 발생할 경우 해당 세그먼트부터 전부 다시 전송하는 방식입니다.
- 한 세그먼트에만 문제가 발생해도 그 후 모든 세그먼트를 다시 재전송해야 한다는 단점이 있습니다.
- 순서 번호 N번에 대한 ACK 세그먼트는 ‘N번만의’ 확인 응답이 아니라 ‘N번까지의’ 확인 응답이라는 점에서, Go-Back-N ARQ의 ACK 세그먼트를 누적 확인 응답(CACK)이라고도 합니다.

![image](https://github.com/user-attachments/assets/d761ca69-5f55-4964-bbb3-9f77e8cc3fa1)

**3) Selective Repeat(SR) ARQ**

- 이름 그대로 누락된 세그먼트에 대해서만 선택적으로 재전송하는 방식입니다.
- 받은 패킷을 재정렬하는 로직이 추가로 필요합니다.
- Go-Back-N ARQ와 마찬가지로 파이프라이닝 방식을 활용해 여러 세그먼트를 전송합니다.
- Go-Back-N ARQ의 ACK 세그먼트가 누적 확인 응답이라면, Selective Repeat ARQ의 ACK 세그먼트는 **개별 확인 응답**입니다.
- 오늘날의 대부분의 호스트는 TCP 통신에서 Selective Repeat ARQ를 지원합니다.

![image](https://github.com/user-attachments/assets/a5953ebd-96e3-4bae-88de-720f77f0bac9)

<br><br>

[참고한 자료]
- <혼자 공부하는 네트워크> - 04-3. TCP의 오류·흐름·혼잡 제어
- https://roughroad.tistory.com/543 - 이미지
